testinfra/utils/ansible_runner.py

```diff
@@ -18,150 +18,189 @@
+            shutil.rmtree(data_dir)
+        if not os.path.exists(data_dir):
+            os.makedirs(data_dir)
+
+        # runner must have an inventory file
+        inv_dir = os.path.join(data_dir, 'inventory')
+        if not os.path.exists(inv_dir):
+            os.makedirs(inv_dir)
+        inv_file = os.path.join(inv_dir, os.path.basename(self.host_list))
+        shutil.copy(self.host_list, inv_file)
+
+        # molecule inventories use lookups
+        this_env = os.environ.copy()
+        env_keys = list(this_env.keys())
+        for ekey in env_keys:
+            if not ekey.startswith('MOLECULE'):
+                this_env.pop(ekey)
+        env_dir = os.path.join(data_dir, 'env')
+        if not os.path.exists(env_dir):
+            os.makedirs(env_dir)
+        env_file = os.path.join(env_dir, 'envvars')
+        with open(env_file, 'w') as f:
+            f.write('---\n')
+            f.write(yaml.dump(this_env))
+
+        # build the kwarg payload ansible-runner requires
+        runner_kwargs = {
+            'private_data_dir': data_dir,
+            'host_pattern': host,
+            'module': module_name,
+            'module_args': module_args,
+            'json_mode' : True if module_name != "shell" else False
+        }
+
+        # ansible-runner does not have kwargs for these
+        for opt in ['become', 'check']:
+            if kwargs.get(opt, False):
+                if 'cmdline' not in runner_kwargs:
+                    runner_kwargs['cmdline'] = '--%s' % opt
+                else:
+                    runner_kwargs['cmdline'] += ' --%s' % opt
+
+        # invoke ansible-runer -> ansible adhoc
+        r = ansible_runner.run(**runner_kwargs)
+
+        # NOTE(ianw): good heavens, strip ascii color, first line and
+        # newlines from something like:
+        #
+        # ---
+        # localhost | CHANGED | rc=0 >>
+        # Linux
+        #
+        # ---
+        #
+        # I am sure this is not the best way...
+        if module_name == "shell":
+            stdout = r.stdout.read()
+            # from https://stackoverflow.com/a/33925425
+            stdout = re.sub(r'(\x9B|\x1B\[)[0-?]*[ -/]*[@-~]', '', stdout)
+            # Ignore everything before the line ending with >> ... you can
+            # get warnings and stuff before that from ansible.
+            stripping_leading = True
+            final_stdout = []
+            for line in stdout.splitlines():
+                if stripping_leading:
+                    if line.endswith('>>'):
+                        stripping_leading = False
+                    continue
+                final_stdout.append(line)
+            # and finally remove any trailing whitespace ...
+            stdout = "\n".join(final_stdout).strip()
+
+            result = {
+                'rc': r.rc,
+                'stdout': stdout,
+                'stderr': None
+            }
+            return result
+        else:
+            return json.load(r.stdout)
```

Comment:
This seems to be pretty close, except ```r.stdout``` doesn't actually seem to consistently contain the full stdout from ansible.  This is especially evident with

```
setup = host.ansible("setup")["ansible_facts"]
````

which returns a very big json list ... it's often cut off.  I think there's something wrong with the chunking done by ansible-runner in the way it copies into stdout.  thus this load just fails on invalid json